File: spring-boot-oauth2-auth-server/src/main/java/com/bittokazi/oauth2/auth/server/app/models/tenant/OauthClient.java
package com.bittokazi.oauth2.auth.server.app.models.tenant;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.nimbusds.jose.shaded.gson.Gson;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.GenericGenerator;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;

@Setter
@Getter
@AllArgsConstructor
@NoArgsConstructor
@Entity
@Table(name = "oauth_client_details")
public class OauthClient implements Serializable  {

    /**
     *
     */
    private static final long serialVersionUID = 1L;

    @Id
    @GeneratedValue(generator = "uuid")
    @GenericGenerator(name = "uuid", strategy = "uuid2")
    private String id;

    @Column(name = "client_id")
    private String clientId;

    @Column(name = "resource_ids")
    private String resourceIds;

    @Column(name = "client_secret")
    @JsonProperty(access = JsonProperty.Access.WRITE_ONLY)
    private String clientSecret;

    @Column(name = "scope")
    private String scope;

    @Column(name = "client_authentication_method")
    private String clientAuthenticationMethod;

    @Column(name = "authorized_grant_types")
    private String authorizedGrantTypes;

    @Column(name = "web_server_redirect_uri")
    private String webServerRedirectUri;

    @Column(name = "authorities")
    private String authorities;

    @Column(name = "access_token_validity")
    private Integer accessTokenValidity;

    @Column(name = "refresh_token_validity")
    private Integer refreshTokenValidity;

    @Column(name = "additional_information")
    private String additionalInformation;

    @Column(name = "require_consent")
    private Boolean requireConsent;

    @Column(name = "post_logout_url")
    private String postLogoutUrl;

    @Column(name = "token_type")
    private String tokenType;

    @Transient
    private Boolean generateSecret = false;

    @Transient
    private String newSecret;

    public Set<String> getScope() {
        try {
            String[] scopes = this.scope.split(",");
            return new HashSet(Arrays.asList(scopes));
        } catch (Exception e) {
            return null;
        }
    }

    public Set<String> getAuthorizedGrantTypes() {
        try {
            String[] grantTypes = this.authorizedGrantTypes.split(",");
            return new HashSet(Arrays.asList(grantTypes));
        } catch (Exception e) {
            return null;
        }
    }

    public Set<String> getWebServerRedirectUri() {
        try {
            String[] webServerRedirectUris = this.webServerRedirectUri.split(",");
            return new HashSet(Arrays.asList(webServerRedirectUris));
        } catch (Exception e) {
            return null;
        }
    }

    public Map<String, Object> getAdditionalInformation() {
        return new Gson().fromJson(this.additionalInformation, HashMap.class);
    }

}

File: spring-boot-oauth2-auth-server/src/main/kotlin/com/bittokazi/oauth2/auth/server/app/models/tenant/OauthClient.kt
package com.bittokazi.oauth2.auth.server.app.models.tenant

import com.fasterxml.jackson.annotation.JsonGetter
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonSetter
import com.nimbusds.jose.shaded.gson.Gson
import jakarta.persistence.*
import org.hibernate.annotations.GenericGenerator
import java.io.Serializable
import java.util.*

@Entity
@JsonIgnoreProperties(ignoreUnknown = true)
@Table(name = "oauth_client_details")
class OauthClient: Serializable {
    @Id
    @GeneratedValue(generator = "uuid")
    @GenericGenerator(name = "uuid", strategy = "uuid2")
    val id: String? = null

    @Column(name = "client_id")
    var clientId: String? = null

    @Column(name = "resource_ids")
    var resourceIds: String? = null

    @Column(name = "client_secret")
    @JsonProperty(access = JsonProperty.Access.WRITE_ONLY)
    var clientSecret: String? = null

    @JsonIgnore
    @Column(name = "scope")
    var scope: String? = null

    @Column(name = "client_authentication_method")
    var clientAuthenticationMethod: String? = null

    @JsonIgnore
    @Column(name = "authorized_grant_types")
    var authorizedGrantTypes: String? = null

    @JsonIgnore
    @Column(name = "web_server_redirect_uri")
    var webServerRedirectUri: String? = null

    @Column(name = "authorities")
    var authorities: String? = null

    @Column(name = "access_token_validity")
    var accessTokenValidity: Int? = null

    @Column(name = "refresh_token_validity")
    var refreshTokenValidity: Int? = null

    @JsonIgnore
    @Column(name = "additional_information")
    var additionalInformation: String? = null

    @Column(name = "require_consent")
    var requireConsent: Boolean? = null

    @Column(name = "post_logout_url")
    var postLogoutUrl: String? = null

    @Column(name = "token_type")
    var tokenType: String? = null

    @Transient
    val generateSecret = false

    @Transient
    var newSecret: String? = null

    fun scopeAsSet(): MutableSet<String> {
        try {
            val scopes = scope!!.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            return mutableSetOf(*scopes)
        } catch (e: Exception) {
            return mutableSetOf()
        }
    }

    @JsonGetter("scope")
    fun getterScope() = scopeAsSet()

    @JsonSetter("scope")
    fun setterScope(data: MutableSet<String>) {
        scope = data.joinToString(",")
    }

    fun authorizedGrantTypesAsSet(): MutableSet<String> {
        try {
            val grantTypes =
                authorizedGrantTypes!!.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            return mutableSetOf(*grantTypes)
        } catch (e: Exception) {
            return mutableSetOf()
        }
    }

    @JsonGetter("authorizedGrantTypes")
    fun getterAuthorizedGrantTypes() = authorizedGrantTypesAsSet()

    @JsonSetter("authorizedGrantTypes")
    fun setterAuthorizedGrantTypes(data: MutableSet<String>) {
        authorizedGrantTypes = data.joinToString(",")
    }

    fun additionalInformationMap(): Map<*, *> = Gson().fromJson(additionalInformation, HashMap::class.java)

    @JsonGetter("additionalInformation")
    fun getterAdditionalInformation() = additionalInformation

    @JsonSetter("additionalInformation")
    fun setterAdditionalInformation(data: String) {
        additionalInformation = data
    }

    fun webServerRedirectUriAsSet(): MutableSet<String>  {
        try {
            val webServerRedirectUris = webServerRedirectUri!!.split(",".toRegex())
                .dropLastWhile { it.isEmpty() }.toTypedArray()
            return when(webServerRedirectUris.size) {
                0 -> mutableSetOf("")
                else -> mutableSetOf(*webServerRedirectUris)
            }
        } catch (e: Exception) {
            return mutableSetOf("")
        }
    }

    @JsonGetter("webServerRedirectUri")
    fun getterWebServerRedirectUri() = webServerRedirectUriAsSet()

    @JsonSetter("webServerRedirectUri")
    fun setterWebServerRedirectUri(data: MutableSet<String>) {
        webServerRedirectUri = data.joinToString(",")
    }

    companion object {
        const val serialVersionUID = 1L
    }
}

